libsel4/tools/rustgenerator.py

```diff
@@ -0,0 +1,615 @@
+                                                             word_size)
+
+    class BitFieldType(Type):
+        """
+        A special 'struct' generated by the bitfield generator
+        """
+
+        def __init__(self, name, size_bits, wordsize):
+            RustGenerator.Type.__init__(self, name, size_bits, wordsize)
+
+        def rust_expression(self, var_name, word_num=0):
+            return "%s.words[%d]" % (var_name, word_num)
+
+    class CapType(Type):
+        """
+        A type that is just a typedef of seL4_CPtr.
+        """
+
+        def __init__(self, name, wordsize):
+            RustGenerator.Type.__init__(self, name, wordsize, wordsize)
+
+    class StructType(Type):
+        """
+        A 'struct' definition.
+        """
+
+        def __init__(self, name, size_bits, wordsize):
+            RustGenerator.Type.__init__(self, name, size_bits, wordsize)
+
+        def rust_expression(self, var_name, word_num, member_name):
+            assert word_num < self.size_bits // self.wordsize
+
+            # Multiword structure.
+            assert self.pass_by_reference()
+            return "(*%s).%s" % (var_name, member_name[word_num])
+
+    class PointerType(Type):
+        """
+        A pointer to a standard type.
+        """
+
+        def __init__(self, base_type, wordsize, mutable):
+            RustGenerator.Type.__init__(self, base_type.name, wordsize, wordsize)
+            self.base_type = base_type
+            self.mutable = mutable
+
+        def render_parameter_name(self, name):
+            if name == "type":
+                name = "type_"
+            if self.mutable:
+                return "%s: *mut %s" % (name, RustGenerator.translate_type(self.name))
+            else:
+                return "%s: *const %s" % (name, RustGenerator.translate_type(self.name))
+
+        def rust_expression(self, var_name, word_num=0):
+            assert word_num == 0
+            return "unsafe { *%s }" % var_name
+
+        def pointer(self):
+            raise NotImplementedError()
+
+    def construction(self, expr, param):
+        if isinstance(param.type, RustGenerator.StructType):
+            return "%s { words: [%s] }" % (param.type.name, expr)
+        else:
+            return "%s as %s" % (expr, RustGenerator.translate_type(param.type.name))
+
+    def init_data_types(self):
+        # TODO refactor, this could be implemented in Generator instead
+        wordsize = self.wordsize
+        BitFieldType = RustGenerator.BitFieldType
+        CapType = RustGenerator.CapType
+        Type = RustGenerator.Type
+        return [
+            # Simple Types
+            Type("int", 32, wordsize),
+            Type("long", wordsize, wordsize),
+
+            Type("seL4_Uint8", 8, wordsize),
+            Type("seL4_Uint16", 16, wordsize),
+            Type("seL4_Uint32", 32, wordsize),
+            Type("seL4_Uint64", 64, wordsize, double_word=(wordsize == 32)),
```

Comment:
Might also be a follow-up task, but it there a way to drop this `double_word` and handle this internally where needed?

Reply:
Yes, `double_word` was set as true when `(size_bits == wordsize*2)`, so I've moved it to `__init__` function of `Type`, and removed this from arguments of the function.